### PR TITLE
GEN-2: updating lock logic for applying deltas

### PR DESCRIPTION
**Pull request by anthonyheidenreich**

### CODE REVIEW

File: database/deltas/execute.go [hunk 1]
```diff
@@ -53,16 +55,23 @@ func Execute(config database.InstanceConfig, schema string, deltas []*Delta) err
...
+		if !locked {
+			return errors.New("unable to acquire lock")
+		}
+		return nil
+	}
+
 	log.Infof("executing %d deltas on %s", len(deltas), schema)
-	// we want to just bail if someone else has the lock
-	locked, err := database.AcquireDatabaseLock(db, LockName, 0)
+	err = net.BackoffExtended(getLock, 10, time.Second, 10*time.Second)
```
Comment (anthonyheidenreich) [inline database/deltas/execute.go:70]: How would you like the timing components passed in?

Comment (apynes) [reply database/deltas/execute.go:70]: Let's put this on the instance config (`gadget/database/model.go (56)`)

File: database/deltas/execute.go [hunk 2]
```diff
@@ -53,16 +55,23 @@ func Execute(config database.InstanceConfig, schema string, deltas []*Delta) err
 	config.Connection = setMultiStatement(config.Connection)
 	db := database.Initialize(&config)
 
+	getLock := func() error {
```
Comment (apynes) [inline database/deltas/execute.go:58]: promote this to a normal function to increase test-ability

**Review by apynes — CHANGES_REQUESTED** [review]

File: database/deltas/execute.go [hunk 3]
```diff
@@ -54,15 +55,11 @@ func Execute(config database.InstanceConfig, schema string, deltas []*Delta) err
 	db := database.Initialize(&config)
 
 	log.Infof("executing %d deltas on %s", len(deltas), schema)
-	// we want to just bail if someone else has the lock
-	locked, err := database.AcquireDatabaseLock(db, LockName, 0)
+	err = net.BackoffExtended(getLock(db), config.NumberOfDeltaLockTries(), config.MinimumWaitBetweenDeltaLockRetries(), config.MaxWaitBetweenDeltaLockRetries())
```
Comment (apynes) [inline database/deltas/execute.go:58]: line too long

File: database/model.go [hunk 4]
```diff
@@ -87,6 +93,27 @@ func (config *InstanceConfig) WaitBetweenRetries() time.Duration {
 	return config.ConnectRetryWait
 }
 
+// NumberOfDeltaLockTries on a connection to the database before failing
+func (config *InstanceConfig) NumberOfDeltaLockTries() int {
+	return config.DeltaLockMaxTries
```
Comment (apynes) [inline database/model.go:98]: default this to like 10 if '<= 0' no retries would not make sense in this context

**Review by apynes — CHANGES_REQUESTED** [review]
2 changes in existing code. We can make this a lot more testable if we add an internal function that Execute calls that allows for injection of the locking function and the database (just make this an interface that satisfies the minimal functional requirements we have here).

File: database/model.go [hunk 5]
```diff
@@ -87,6 +93,30 @@ func (config *InstanceConfig) WaitBetweenRetries() time.Duration {
 	return config.ConnectRetryWait
 }
 
+// NumberOfDeltaLockTries on a connection to the database before failing
+func (config *InstanceConfig) NumberOfDeltaLockTries() int {
+	if config.DeltaLockMaxTries < 10 {
+		return 10
+	}
+	return config.DeltaLockMaxTries
```
Comment (apynes) [inline database/model.go:101]:
```suggestion
	if config.DeltaLockMaxTries == 0 {
		config.DeltaLockMaxTries = DefaultMaxTries
	}
	return config.DeltaLockMaxTries
```

Comment (apynes) [reply database/model.go:101]: And add DefaultMaxTries as a constant

**Review by apynes — CHANGES_REQUESTED** [review]
Tests? You could create the database interface and add tests as a separate ticket if you want, just drop a todo and a jira ticket

**Comment by nmarie22** [comment]
@apynes see most recent commit. The database package functions won't take a fake database as an argument. What am I missing?

**Comment by nmarie22** [comment]
@apynes tests added

**Review by apynes — APPROVED** [review]
LGTM :t-rex:

**Pull request merged**